File: Leetcode/lcopn.py
# leetcode no.17 Letter Combinations of a Phone Number

class Solution:
    def letterCombinations(self, digits: str):
        if len(digits) == 0:
            return []
        dial = [[] for _ in range(10)]
        dial[2] = ["a", "b", "c"]
        dial[3] = ["d", "e", "f"]
        dial[4] = ["g", "h", "i"]
        dial[5] = ["j", "k", "l"]
        dial[6] = ["m", "n", "o"]
        dial[7] = ["p", "q", "r", "s"]
        dial[8] = ["t", "u", "v"]
        dial[9] = ["w", "x", "y", "z"]
        s = set([])

        def dfs(idx, pattern):
            if idx == len(digits):
                s.add(pattern)
                return
            for c in dial[int(digits[idx])]:
                dfs(idx + 1, pattern + c)

        dfs(0, "")
        ret = list(s)
        return ret


sol = Solution()
print(sol.letterCombinations("23"))

File: Leetcode/lcopn.cpp
// leetcode no.17 Letter Combinations of a Phone Number

#include <iostream>
#include <string>
#include <vector>

using namespace std;


class Solution {
public:
    vector<string> letterCombinations(string digits) {
        vector<string> dial = {"", "", "abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"};
        vector<string> store = {""};
		for(int i = 0; i < digits.length(); i++){
        	int index = digits[i] - '0';
        	vector<string> temp;
			temp.assign(store.begin(), store.end());
			store.clear();
			for(auto& elem : temp){
				for(int j = 0; j < dial[index].length(); j++){
					string c = dial[index].substr(j, 1);
					store.push_back(elem + c);
				}
			}
		}
		return store;
    }
};

int main(void){
	string input = "23";
	Solution sol;
	vector<string> result = sol.letterCombinations(input);
	for(auto& elem: result){
		cout << elem << " ";
	}
	cout << "\n";
	return 0;
}
